pkg/build/controller/image_change_controller.go

```diff
@@ -0,0 +1,72 @@
+package controller
+
+import (
+	"github.com/golang/glog"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client/cache"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	imageapi "github.com/openshift/origin/pkg/image/api"
+	osclient "github.com/openshift/origin/pkg/client"
+
+)
+
+// ImageChangeController watches for changes to ImageRepositories and triggers
+// builds when a new version of a tag referenced by a BuildConfig
+// is available.
+type ImageChangeController struct {
+	Client *osclient.Client
+	NextImageRepository  func() *imageapi.ImageRepository
+	BuildConfigStore     cache.Store
+	BuildCreator buildCreator
+	// Stop is an optional channel that controls when the controller exits
+	Stop <-chan struct{}
+}
+
+type buildCreator interface {
+	CreateBuild(build *buildapi.BuildConfig, imageSubstitutions map[string]string) error
+}
+
+// Run processes ImageRepository events one by one.
+func (c *ImageChangeController) Run() {
+	go util.Until(c.HandleImageRepo, 0, c.Stop)
+}
+
+// HandleImageRepo processes the next ImageRepository event.
+func (c *ImageChangeController) HandleImageRepo() {
+	glog.V(4).Infof("Waiting for imagerepo change")
+	imageRepo := c.NextImageRepository()
+	glog.V(4).Infof("Build image change controller detected imagerepo change %s", imageRepo.DockerImageRepository)
+	firedTriggerForConfig := make(map[string]*buildapi.ImageChangeTrigger)
+	imageSubstitutions:=make(map[string]string)
+	
+	for _, bc := range c.BuildConfigStore.List() {
+		config := bc.(*buildapi.BuildConfig)
+		glog.V(4).Infof("Detecting changed images for buildConfig %s", config.Name)
+
+		// Extract relevant triggers for this imageRepo for this config
+		var triggerForConfig *buildapi.ImageChangeTrigger
+		for _, trigger := range config.Triggers {
+			// for every ImageChange trigger, record the image it substitutes for and get the latest
+			// image id from the imagerepository.  We will substitute all images in the buildconfig
+			// with the latest values from the imagerepositories.
+			if trigger.Type == buildapi.ImageChangeType {
+				if repoImageID, repoHasTag := imageRepo.Tags[trigger.ImageChange.Tag]; repoHasTag {
+					imageSubstitutions[trigger.ImageChange.Image]=imageRepo.DockerImageRepository+":"+repoImageID
```

Comment:
I misanswered this on IRC, I thought you were referring to the other direction.  ImageRepositories listen for any tag being pushed.  That Tags map[] is not defined in the json, it gets populated as images are pushed to the registry.